sql/analyzer/resolve_subqueries.go

```diff
@@ -110,47 +105,80 @@ func flattenTableAliases(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope
+		// Specifically, if the subquery isn't resolved or a column can't be found in the scope node, wait until a later pass.
+		if !finalize && (ErrValidationResolved.Is(err) || sql.ErrTableColumnNotFound.Is(err) || sql.ErrColumnNotFound.Is(err)) {
+			// keep the work we have and defer remainder of analysis of this subquery until a later pass
+			return sq.WithQuery(analyzed), transform.NewTree, nil
+		}
+		return nil, transform.SameTree, err
+	}
 
-func finalizeSubqueryExpressions(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
-	return resolveSubqueryExpressionsHelper(ctx, a, n, scope, SelectAllBatches, sel)
+	//todo(max): Infinite cycles with subqueries, unions, ctes, catalog.
+	// we squashed most negative errors, where a rule fails to report a plan change
+	// to the expense of positive errors, where a rule reports a change when the plan
+	// is the same before/after.
+	// .Resolved() might be useful for fixing these bugs.
+	return sq.WithQuery(StripPassthroughNodes(analyzed)), transform.NewTree, nil
 }
 
-func resolveSubqueryExpressionsHelper(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, batchSel BatchSelector, ruleSel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
-	return transform.NodeExprsWithNode(n, func(n sql.Node, e sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
-		s, ok := e.(*plan.Subquery)
-		// We always analyze subquery expressions even if they are resolved, since other transformations to the surrounding
-		// query might cause them to need to shift their field indexes.
-		if !ok {
-			return e, transform.SameTree, nil
-		}
+// analyzeSubqueryAlias runs analysis on the specified subquery alias, |sqa|. The |finalize| parameter indicates if this is
+// the final run of the analyzer on the query before execution, which means all rules, starting from the default-rules
+// batch are processed, otherwise only the once-before-default batch of rules is processed for all other non-final passes.
+func analyzeSubqueryAlias(ctx *sql.Context, a *Analyzer, sqa *plan.SubqueryAlias, scope *Scope, sel RuleSelector, finalize bool) (sql.Node, transform.TreeIdentity, error) {
+	subScope := newScopeWithDepth(scope.RecursionDepth() + 1)
+	if scope != nil && len(scope.nodes) > 0 {
+		// As of MySQL 8.0.14, MySQL provides OUTER scope visibility to derived tables. Unlike LATERAL scope visibility, which
+		// gives a derived table visibility to the adjacent expressions where the subquery is defined, OUTER scope visibility
+		// gives a derived table visibility to the OUTER scope where the subquery is defined.
+		// https://dev.mysql.com/blog-archive/supporting-all-kinds-of-outer-references-in-derived-tables-lateral-or-not/
+		// We don't include the current inner node so that the outer scope nodes are still present, but not the lateral nodes
+		subScope.nodes = scope.InnerToOuter()
+		sqa.OuterScopeVisibility = true
```

Comment:
might be worth considering some alternatives that expose information that more broadly applicable:
- relational subquery goes one step further to determine if it's correlated with an outer scope. you would only know this information after resolving columns.
- relation subquery has `outerCols` / `outerColsReferenced` / etc to indicate both that it is correlated, and which outer field parametrize this subquery

Reply:
we want a stricter definition of cacheable depending on whether a relational subquery is correlated, not just that there is an outer scope. this info is reusable for ex: decorrelating subqueries, but also for queries like this:
```SQL
> explain select u from uv where u in (select * from (select x+1 from xy where x = u) s);
```
the above query currently panics, because we fail to realize that the relational subquery inside of the scalar subquery is correlated with outer scope, and try applying the `IndexedInSubqueryFilter` -- an invalid transformation if the subquery is not cacheable. It is important not to break similar but valid queries like the one below, which turbine depends on:
```SQL
> explain select u from uv where u in (select u+1 from uv);
+-----------------------------------------------+
| plan                                          |
+-----------------------------------------------+
| Project                                       |
|  ├─ columns: [uv.u]                           |
|  └─ IndexedInSubqueryFilter(uv.u IN ((Project |
|      ├─ columns: [(uv.u + 1)]                 |
|      └─ Table(uv)                             |
|          └─ columns: [u]                      |
|     )))                                       |
|      └─ IndexedTableAccess(uv)                |
|          └─ index: [uv.u]                     |
+-----------------------------------------------+
```

Reply:
The stricter property would be recursive, also, root scope -> scalar SQ -> rel SQ -> scalar SQ -> rel SQ ... -> rel SQ with root scope reference

Reply:
I added `SubqueryAlias.CanCacheResults`, to match `Subquery.CanCacheResults`. That gets us back to the same caching behavior as before, although I did find a few places where we were previously caching a derived table that shouldn't have been cached (e.g. a derived table that had a `RAND()` expression in it was previously cached, but now is not eligible for caching because of the non-determinate expression). 

I just tried out both of those queries you suggested and confirmed they both execute correctly on GMS now and match MySQL's behavior. Happy to make sure those are included in the test suite this afternoon, too.